### DISCUSSION TITLE
Changing addresses for individual potentiometers

### DISCUSSION THREAD

**Discussion by JJPeachum** [tttapa/Control-Surface#1052] [Q&A]
Hi there!
I'm trying to build a MIDI-controller with 6 potentiometers sending CC-values. In addition I would like to use 4 buttons to change the MIDI-addresses of each potentiometer individually: 2 buttons (left-right) to select a potentiometer, 2 buttons (up-down) to change the address of the selected potentiometer.

I'm quite new to more complex Arduino projects and the control_surface-library and was not able to find a solution. Using a variable for the address wasn't working and seems not to be the right way of doing it. I've read through a lot of forums and the FAQs but so far I was only able to change the address of one potentiometer using banks:

```
#include <Control_Surface.h>

USBDebugMIDI_Interface midi = 115200;

Bank<127> bank(1);

IncrementDecrementSelector<127> selector{
  bank,
  {7, 8},
  Wrap::Wrap,
};

Bankable::CCPotentiometer pot1 = {
  {bank, BankType::CHANGE_ADDRESS},
  A0, {0, CHANNEL_1}
  };

void setup() { 
  Control_Surface.begin(); 
}

void loop() { 
  Control_Surface.loop(); 
}
```

I really appreciate any help and would love to learn more. Thanks in advance!

**Comment by tttapa** [comment]
You can either use `bank.select(n);`, or use `CCPotentiometer::setAddress(...);`  
https://tttapa.github.io/Control-Surface-doc/Doxygen/db/d32/classCCPotentiometer.html#a047faf611c45404fcc29b998f3d3f47a.

**Comment by JJPeachum** [comment]
Thank you for the fast replay! But unfortunately I don't know how to integrate it in the code.

**Reply by tttapa** [reply]
Start from the [Control-Change-Potentiometer example](https://tttapa.github.io/Control-Surface-doc/Doxygen/de/de4/Control-Change-Potentiometer_8ino-example.html).

Then add your logic for reading the up/down/left/right buttons. Finally, change the address of the appropriate potentiometer. For example:

```cpp
if (/* some condition based on the buttons */) {
  potentiometer.setAddress(0x40); // set the controller number
}
if (/* some condition based on the buttons */) {
  potentiometer.setAddress(potentiometer.getAddress() + 1); // increment the controller number
}
```

**Reply by JJPeachum** [reply]
I was just trying it out and got the following error:
 
> Compilation error: 'class CS::CCPotentiometer' has no member named 'setAddress'; did you mean 'address'?

My code looked like this now:

```
#include <Control_Surface.h> // Include the Control Surface library

const int button_up = 8;
const int button_down = 7;
const int button_left = 10;
const int button_right = 9;

char button_upState = 0;
char button_downState = 0;
char button_leftState = 0;
char button_rightState = 0;

// Instantiate a MIDI over USB interface.
USBDebugMIDI_Interface midi;

// Instantiate a CCPotentiometer object
CCPotentiometer potentiometer {
  A0,             // Analog pin connected to potentiometer
  {0, CHANNEL_1}, // Channel volume of channel 1
};

void setup() {
  pinMode(button_up, INPUT_PULLUP);
  pinMode(button_down, INPUT_PULLUP);
  pinMode(button_left, INPUT_PULLUP);
  pinMode(button_right, INPUT_PULLUP);

  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface

  button_upState = digitalRead(button_up);
  button_downState = digitalRead(button_down);
  button_leftState = digitalRead(button_left);
  button_rightState = digitalRead(button_right);

  if (button_rightState == LOW) {
    potentiometer.setAddress(0x40); // set the controller number
  }
  if (button_upState == LOW) {
    potentiometer.setAddress(potentiometer.getAddress() + 1); // increment the controller number
  }
}
```

(Ignore the use of the buttons in the if-condition – I simply wanted to figure out what they are doing now…)

**Reply by tttapa** [reply]
Are you using the latest `main` version of Control Surface?

**Reply by JJPeachum** [reply]
Oh, that's it! I've downloaded the latest version now and it works perfectly. Thank you so much!
